### PR TITLE
emit namespaces for top-level extern symbols.

### PR DESCRIPTION
**Pull request by rkirov**
skip platform based externs as they might mirror types in lib.d.ts.

### CODE REVIEW

File: src/main/java/com/google/javascript/clutz/DeclarationGenerator.java [hunk 1]
```diff
@@ -223,15 +224,44 @@ public String produceDts(Compiler compiler, Depgraph depgraph) {
         int lastDot = symbol.getName().lastIndexOf('.');
         namespace = lastDot >= 0 ? symbol.getName().substring(0, lastDot) : "";
       }
-      declareNamespace(namespace, symbol, isDefault, compiler, transitiveProvides);
+      int valueSymbolsWalked = declareNamespace(namespace, symbol, isDefault, compiler, transitiveProvides);
+
+      // skip emitting goog.require declarations for value empty namespaces, as calling typeof
+      // does not work them.
```
Comment (mprobst) [inline src/main/java/com/google/javascript/clutz/DeclarationGenerator.java]: nit: does not work _for_ them?

File: src/main/java/com/google/javascript/clutz/DeclarationGenerator.java [hunk 2]
```diff
@@ -223,15 +224,44 @@ public String produceDts(Compiler compiler, Depgraph depgraph) {
...
+    // symbols, even though they have no explicit provides.
+    for (TypedVar symbol : topScope.getAllSymbols()) {
+      CompilerInput symbolInput = compiler.getInput(new InputId(symbol.getInputName()));
+      if (symbolInput == null || !symbolInput.isExtern()) continue;
+      if (isPlatformExtern(symbolInput)) continue;
+      // Only emitting top-level symbols, since they transitively contain all children.
+      if (symbol.getName().contains(".")) continue;
+      declareNamespace(symbol.getName(), symbol, false, compiler, transitiveProvides);
+      // we do not declare modules or goog.require support, because externs types should not be
+      // visible by TS code.
```
Comment (mprobst) [inline src/main/java/com/google/javascript/clutz/DeclarationGenerator.java]: uber-nit: visible _from_ TS code?

File: src/resources/third_party_externs.js [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+angular.Scope.prototype.$$phase;
+
+/**
+ * @param {(string|function(!angular.Scope))=} opt_exp
+ * @return {*}
+ */
+angular.Scope.prototype.$apply = function(opt_exp) {};
+
+/**
+ * @param {(string|function(!angular.Scope))=} opt_exp
```
Comment (mprobst) [inline src/resources/third_party_externs.js]: what if one of these APIs mentions e.g. `Element`? I think it doesn't happen in the Angular externs, and even if should mostly just work, but might be worth thinking about and/or adding a test.

File: src/test/java/com/google/javascript/clutz/ProgramSubject.java [hunk 4]
```diff
@@ -108,8 +110,11 @@ private String parse() throws AssertionError {
       roots.add("main.js");
     }
 
+    List<SourceFile> externs = withExterns ? DeclarationGenerator.getDefaultExterns(opts) : NO_EXTERNS;
```
Comment (mprobst) [inline src/test/java/com/google/javascript/clutz/ProgramSubject.java]: nit: ternary followed by condition is somewhat odd, could you just fold the ternary into the conditional?

**Comment by mprobst** [comment]
LGTM.

Nice!

File: src/resources/third_party_externs.js (hunk 3 above)
Comment (rkirov) [inline src/resources/third_party_externs.js]: added bootstrap.

File: src/test/java/com/google/javascript/clutz/ProgramSubject.java (hunk 4 above)
Comment (rkirov) [inline src/test/java/com/google/javascript/clutz/ProgramSubject.java]: done.

**Pull request merged**